doc/xrcvd/transceiver-monitor-hld.md

```diff
@@ -283,6 +282,62 @@ An explicit "error clear event" can be added if some vendor's platform supports
+
+- Main process works as a "Subject", it subscribes CONFIG_DB PORT table change and publish port change event.
+- State machine process works as a "Observer", it subscribes port change event and update local port mapping accordingly.
+- DOM sensor update thread works as a "Observer" either, it subscribes port change event and update local port mapping accordingly.
+
+Port change event contains following data:
+
+- ASIC index which indicates the DB namespace.
+- Logical port name. Get from the key of PORT table. E.g, for key "PORT|Ethernet0", the logical port name is "Ethernet0".
+- Physic port index. Get from "index" field of PORT table.
+- Event type. Can be add or remove.
+
+As port mapping information might be updated during runtime, the global port mapping information cannot be shared properly among main process, state machine process and DOM sensor update thread. A possible solution is to use share memory between different processes, but it will introduce process level lock to many places which is hard to maintain. So, a simple solution is to store local port mapping information in main process, state machine process and DOM sensor update thread and update them via event. In this case, no explicit lock is needed and we can keep the logic as simple as it is. Of course it takes more memory, but port mapping information would be very small which should not cause any memory issue.
+
+##### 1.4.3.1 Subscribe CONFIG_DB PORT table change #####
+
+SONiC has implemented a way to "select" data changes from redis. xcvrd should reuse this "select" infrastructure to listen CONFIG_DB PORT table change. The workflow is like:
+
+1. For each ASIC namespace, create a selectable object which point to CONFIG_DB PORT table
+2. Add each selectable object to the select queue
+3. Select DB event in a while loop
+4. If there is any selectable object ready for read, check:
+
+    - Entry added, trigger a port change event with event type "Add"
+    - Entry removed, trigger a port change event with event type "Remove"
+    - Entry updated, skip
+
+5. Notify observers
+
+##### 1.4.3.2 Handle port change event in state machine process #####
+
+As state machine task is a process, it should use multiprocessing.Queue to receive the port change event. Once a port change event arrives, it should update local port mapping information first, and if it is a remove event, state machine task should remove transceiver information from table TRANSCEIVER_INFO, TRANSCEIVER_STATUS and TRANSCEIVER_DOM_SENSOR; if it is an add event, there could be 4 cases:
+
+- Transceiver information is already in DB which means that a logical port with the same physical index is already in DB. Copy the data from DB and create a new entry to table TRANSCEIVER_DOM_INFO, TRANSCEIVER_STATUS_INFO and TRANSCEIVER_INFO whose key is the newly added logical port name.

```

Comment:
with the same physical index is "already in DB"   -----> "already exists"      to avoid repeat